servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleThreadedExecutor.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.ref.WeakReference;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+/**
+ * Executor that can execute one task at a time on a dedicated worker thread. While concurrent tasks are generally
+ * rejected, the `execute(Runnable)` method provides special behavior if the invoking thread is the executor worker
+ * thread: the `Runnable` will be executed immediately on the calling thread.
+ */
+class SingleThreadedExecutor implements java.util.concurrent.Executor, AutoCloseable {
+    static final Logger LOGGER = LoggerFactory.getLogger(SingleThreadedExecutor.class);
+    private static final String DEFAULT_NAME_PREFIX = "servicetalk-solo";
+    private static final AtomicInteger INSTANCE_COUNT = new AtomicInteger();
+    final boolean nonblocking;
+    final boolean interruptOnClose;
+    private final WorkerThread thread;
+
+    SingleThreadedExecutor() {
+        this(DEFAULT_NAME_PREFIX);
+    }
+
+    SingleThreadedExecutor(String namePrefix) {
+        this(namePrefix, true, true);
+    }
+
+    SingleThreadedExecutor(String namePrefix, boolean nonblocking, boolean interruptOnClose) {
+        String name = namePrefix + "-" + INSTANCE_COUNT.incrementAndGet();
+        this.nonblocking = nonblocking;
+        this.interruptOnClose = interruptOnClose;
+        thread = new WorkerThread(this, name);
+        thread.start();
+    }
+
+    @Override
+    public void execute(final Runnable command) {
+        if (Thread.currentThread() == thread) {
+            try {
+                command.run();
+            } catch (Throwable all) {
+                LOGGER.warn("Uncaught throwable from command " + command, all);
+            }
+        } else {
+            try {
+                if (nonblocking) {
+                    if (!thread.queue.offer(command, 100, TimeUnit.MICROSECONDS)) {
+                        throw new RejectedExecutionException("Executor is busy");
```

Comment:
I'd definitely like to see some hints as a user why this rejection happened. Perhaps adding the hint about the `nonblocking` flag being true to the message would suffice?